### PR TITLE
Add support for messaging with strong auth

### PR DESCRIPTION
**Pull request by Wnt**
## After this change

- Database stores sensitivity flag of messages
- Employee messaging API's MessageThread draft and MessageThread PUT and POST endpoints support setting sensitivity flag
- If MessageThread is set as sensitive and sender account type is not `PERSONAL`, message contains multiple recipients or recipien type != `CHILD` API endpoint POST createMessage will return a BadRequest error
- Citizen messaging API's MessageThread GET returns either Redacted or Regular versions of messages marked as sensitive depending on the auth level
- Redacted doesn't contain `MessageList` and contains server-side computed fields `lastMessageSentAt` and `hasUnreadMessages`
- Citizen front-end data model and front-end components support this new data model: `type CitizenMessageThread = Redacted | Regular`

## Still left to do for end-to-end support of strong auth messaging
- Add support in Employee front-end for creating sensitive messages threads
- Add supports in Citizen front-end for displaying redacted version of message threads
- Add indicators to Citizen front-end for sensitive message threads

### CODE REVIEW

File: service/src/main/resources/db/migration/R__message_thread_sensitivity_flag.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+-- TODO rename this file into V###__message_thread_sensitivity_flag.sql and add to migrations.txt
```
Comment (Joosakur) [inline service/src/main/resources/db/migration/R__message_thread_sensitivity_flag.sql:1]: todo

File: service/src/main/kotlin/fi/espoo/evaka/messaging/Message.kt [hunk 2]
```diff
@@ -48,16 +50,52 @@ data class MessageThreadStub(
...
+    MESSAGE_THREAD
+}
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+sealed class CitizenMessageThread(val type: MessageThreadType) {
+    abstract val id: MessageThreadId
+    abstract val urgent: Boolean
+    abstract val children: List<MessageChild>
+
+    @JsonTypeName("REDACTED_MESSAGE_THREAD")
```
Comment (Joosakur) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/Message.kt:68]: to be replaced with The new way (TM)

File: frontend/src/citizen-frontend/messages/state.tsx [hunk 3]
```diff
@@ -59,20 +59,29 @@ const defaultState: MessagePageState = {
...
-          }))
-        }
-      : t
+export const isRedactedThread = (
+  thread: CitizenMessageThread
+): thread is CitizenMessageThread.Redacted =>
+  thread.type === 'REDACTED_MESSAGE_THREAD'
+export const isRegularThread = (
+  thread: CitizenMessageThread
+): thread is CitizenMessageThread.Regular => thread.type === 'MESSAGE_THREAD'
```
Comment (Joosakur) [inline frontend/src/citizen-frontend/messages/state.tsx:68]: note to self: codegen could maybe be extended to generate these automatically for all sealed classes

**Review by Joosakur — COMMENTED** [review]
Looks good to me!

File: service/src/main/kotlin/fi/espoo/evaka/messaging/Message.kt (hunk 2 above)
Comment (Wnt) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/Message.kt:68]: I'll test this if it works without the proposed typed Paged system!

File: service/src/main/kotlin/fi/espoo/evaka/messaging/MessageControllerCitizen.kt [hunk 4]
```diff
@@ -88,25 +89,60 @@ class MessageControllerCitizen(
...
+    }
+
+    fun redactOrWrapMessageThread(
+        user: AuthenticatedUser.Citizen,
+        messageThread: MessageThread
+    ): CitizenMessageThread {
+        if (user.authLevel != CitizenAuthLevel.STRONG && messageThread.sensitive)
+            return redactMessageTread(user, messageThread)
+
+        return CitizenMessageThread.Regular(
```
Comment (Wnt) [inline service/src/main/kotlin/fi/espoo/evaka/messaging/MessageControllerCitizen.kt:114]: Move to CitizenMessageThread.Regular.from()...

File: frontend/src/lib-common/api-types/messaging.ts [hunk 5]
```diff
@@ -46,13 +47,31 @@ export const deserializeMessage = (message: JsonOf<Message>): Message => ({
...
+  messages.map((m) => deserializeMessage(m))
+
 export const deserializeMessageThread = (
   json: JsonOf<MessageThread>
 ): MessageThread => ({
   ...json,
   messages: json.messages.map(deserializeMessage)
 })
 
+export const deserializeCitizenMessageThread = (
```
Comment (Wnt) [inline frontend/src/lib-common/api-types/messaging.ts:60]: Move to citizen

File: frontend/src/lib-customizations/defaults/citizen/i18n/sv.tsx [hunk 6]
```diff
@@ -457,7 +457,9 @@ const sv: Translations = {
       text: 'Meddelanden kan inte återställas efter radering.',
       cancel: 'Radera inte',
       confirm: 'Radera tråden'
-    }
+    },
+    sensitive: 'Känslig meddelandetråd',
```
Comment (Wnt) [inline frontend/src/lib-customizations/defaults/citizen/i18n/sv.tsx:461]: Double check translations from Espoo

File: service/src/main/kotlin/fi/espoo/evaka/messaging/Message.kt (hunk 2 above)
Comment (Wnt) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/Message.kt:68]: @akheron tested this and this still would need the typed Paged system in order to remove the enum + type field + type parameter.

File: service/src/main/kotlin/fi/espoo/evaka/messaging/MessageControllerCitizen.kt (hunk 4 above)
Comment (Wnt) [reply service/src/main/kotlin/fi/espoo/evaka/messaging/MessageControllerCitizen.kt]: Done

File: service/src/main/resources/db/migration/R__message_thread_sensitivity_flag.sql (hunk 1 above)
Comment (Wnt) [reply service/src/main/resources/db/migration/R__message_thread_sensitivity_flag.sql]: Done

**Pull request merged**